File: go/minimum-add-to-make-parentheses-valid.go
func minAddToMakeValid(s string) int {
    balance := 0
    add := 0
    for i := 0; i < len(s); i++ {
        if s[i] == '(' {
            balance += 1
        }
        if s[i] == ')' {
            balance -= 1
        }
        if balance < 0 {
            balance = 0
            add += 1
        }
    }
    add += balance
    return add
}

File: python/minimum-add-to-make-parentheses-valid.py
class Solution:
    def minAddToMakeValid(self, s: str) -> int:
        balance = 0
        additional_chars = 0
        for c in s:
            if c == "(":
                balance += 1
            if c == ")":
                balance -= 1
            if balance < 0:
                additional_chars += 1
                balance = 0
        additional_chars += balance
        return additional_chars
